# 5_5mC

This document describes a minimal workflow to run **metylation analysis**.


## Overview
1. Tool installation
2. Basecalling
3. Generating BED files
4. Calculating DNA methylation levels
5. Figure generation
 5.1 Methylation levels by TE class and genomic region
 5.2 Methylation levels of conserved genes vs. other genes
 5.3 Scatter plot of methylation level vs. gene expression

## 1. Tool installation
```bash
cd 5mC_wd
conda activate ntLink

#bedファイル作成用ツールのインストール
conda install -c bioconda bedops
conda install -c bioconda bedtools

#doradoは金先生がダウンロードしたもの(?)がパスが通っていて使えたので、そのまま使用しており自分ではダウンロードしていません。
#dorado -v
#0.7.1

#modkitのインストール
wget https://github.com/nanoporetech/modkit/releases/download/v0.6.0/modkit_v0.6.0_u16_x86_64.tar.gz
tar -xzvf modkit_v0.6.0_u16_x86_64.tar.gz
```

## 2. Basecalling
```bash
dorado download --model dna_r9.4.1_e8_sup@v3.3

dorado basecaller dna_r9.4.1_e8_sup@v3.3 \
  /data3/Sequence/Nanopore/FC1/1st/pod5_pass/ \
  --modified-bases 5mCG_5hmCG \
  --reference uc.FINAL.fasta \
  > dorado_basecalling.bam

samtools sort -@ 100 -m 10G ./dorado_basecalling.bam -o ./sorted_dorado_basecalling.bam
samtools index  ./sorted_dorado_basecalling.bam

#メモ：basecallingのサマリー#
./dist_modkit_v0.6.0_68b540b/modkit summary ./sorted_dorado_basecalling.bam
> sampling 10042 reads from BAM
> calculating threshold at 10(th) percentile
> calculated thresholds: C: 0.734375
# bases             C
# total_reads_used  10042
# count_reads_C     10042
# pass_threshold_C  0.734375
 base  code  pass_count  pass_frac    all_count  all_frac
 C     -     1542396     0.78524804   1652930    0.75826097
 C     h     42697       0.021737438  84901      0.03894727
 C     m     379122      0.19301452   442065     0.20279178
#修飾確率0.734375以上で5mCと判定される。
```

## 3. Generating BED files
```bash
#exon, intron, gene, promoter, intergenic, repeatのbedファイル作成
#intronの復元
gtf2bed < jnig15_withGENE_withCodons.gtf > uc.FINAL.bed

#exon,intron,geneのbed作成
awk '$8=="gene"   {OFS="\t"; print $1,$2,$3,$4,$5,$6}' uc.FINAL.bed > gene.bed
awk '$8=="exon"   {OFS="\t"; print $1,$2,$3,$4,$5,$6}' uc.FINAL.bed > exon.bed
awk '$8=="intron" {OFS="\t"; print $1,$2,$3,$4,$5,$6}' uc.FINAL.bed > intron.bed

# promoter（両鎖の5' upstream 2kb）のbed作成
samtools faidx uc.FINAL.fasta
bedtools flank -i gene.bed -g uc.FINAL.fasta.fai -l 2000 -r 0 -s \
  | bedtools intersect -a - -b gene.bed -s -v \
  > promoter.bed

# intergenic(exon, intron, promoter, gene以外の領域)のbed作成
cut -f1,2 uc.FINAL.fasta.fai | sort -k1,1 > genome.sizes
cat exon.bed intron.bed promoter.bed gene.bed > in.bed
awk '$6=="+"' in.bed | cut -f1-3 | bedtools sort -i - | bedtools merge -i - > plus.bed
awk '$6=="-"' in.bed | cut -f1-3 | bedtools sort -i - | bedtools merge -i - > minus.bed
bedtools complement -i plus.bed  -g genome.sizes > plus.complement.3col.bed
bedtools complement -i minus.bed -g genome.sizes > minus.complement.3col.bed
awk 'BEGIN{OFS="\t"}{print $1,$2,$3,".",".","+"}' plus.complement.3col.bed  > plus.complement.bed
awk 'BEGIN{OFS="\t"}{print $1,$2,$3,".",".","-"}' minus.complement.3col.bed > minus.complement.bed
cat plus.complement.bed minus.complement.bed | bedtools sort -i - > intergenic.bed
rm in.bed plus.bed minus.bed plus.complement.3col.bed minus.complement.3col.bed plus.complement.bed minus.complement.bed

#repeat配列のbed
cp /data2/nojiri/earlgrey_wd/JNIGearlGreyOutputs/JNIG_EarlGrey/JNIG_summaryFiles/JNIG.filteredRepeats.bed ./
```

## 4. Calculating DNA methylation levels
```bash
./dist_modkit_v0.6.0_68b540b/modkit pileup --cpg --ref ./uc.FINAL.fasta --modified-bases 5mC --threads 100 ./sorted_dorado_basecalling.bam ./dorado_modkit.bed
sort -k1,1 -k2,2n ./dorado_modkit.bed > dorado_modkit.sorted.bed
bgzip -c dorado_modkit.sorted.bed > dorado_modkit.bed.gz
tabix -p bed dorado_modkit.bed.gz

./dist_modkit_v0.6.0_68b540b/modkit stats dorado_modkit.bed.gz \
  --regions exon.bed \
  -o 5mC_in_exon.tsv \
  --mod-codes m

./dist_modkit_v0.6.0_68b540b/modkit stats dorado_modkit.bed.gz \
  --regions intron.bed \
  -o 5mC_in_intron.tsv \
  --mod-codes m
  
./dist_modkit_v0.6.0_68b540b/modkit stats dorado_modkit.bed.gz \
  --regions gene.bed \
  -o 5mC_in_gene.tsv \
  --mod-codes m

./dist_modkit_v0.6.0_68b540b/modkit stats dorado_modkit.bed.gz \
  --regions promoter.bed \
  -o 5mC_in_promoter.tsv \
  --mod-codes m

./dist_modkit_v0.6.0_68b540b/modkit stats dorado_modkit.bed.gz \
  --regions intergenic.bed \
  -o 5mC_in_intergenic.tsv \
  --mod-codes m

./dist_modkit_v0.6.0_68b540b/modkit stats dorado_modkit.bed.gz \
  --regions JNIG.filteredRepeats.bed \
  -o 5mC_in_repeat.tsv \
  --mod-codes m

awk '$8 >= 60 {print $4}' 5mC_in_gene.tsv > methylated_gene.txt
```

## 5. Figure generation
### 5.1 Methylation levels by TE class and genomic region
```python
python3
#violin plot by TE category
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("5mC_in_repeat.tsv", sep="\t")

#メモ：TEのアノテーション名#
>>> df["name"].unique()
array(['Unknown', 'Simple_repeat', 'LINE/Penelope', 'Low_complexity',
       'DNA/CMC-Transib', 'LINE/L2', 'DNA/CMC-EnSpm', 'DNA/hAT-hATm',
       'DNA/hAT-Tip100', 'DNA/CMC-Chapaev', 'DNA/hAT-Ac', 'DNA',
       'DNA/TcMar-ISRm11', 'DNA/hAT-hATx', 'DNA/TcMar-Fot1',
       'LINE/L1-Tx1', 'DNA/TcMar-Tc1', 'Satellite', 'DNA/Kolobok-Hydra',
       'LINE/CR1', 'RC/Helitron', 'DNA/hAT-Tag1', 'DNA/P',
       'DNA/PIF-ISL2EU', 'LINE', 'DNA/Sola-2', 'DNA/Sola-1',
       'DNA/hAT-hAT5', 'LINE/RTE-BovB', 'DNA/PiggyBac', 'PLE/Chlamys',
       'LINE/Proto2', 'DNA/hAT-Blackjack', 'LTR/Gypsy',
       'DNA/TcMar-Mariner', 'DNA/PIF-Harbinger', 'PLE/Hydra',
       'DNA/hAT-Charlie', 'LTR/Pao', 'DNA/Zator', 'DNA/hAT-hATw',
       'DNA/Maverick', 'DNA/CMC-Chapaev-3', 'DNA/Ginger-1',
       'DNA/hAT-Pegasus', 'DNA/MULE-MuDR', 'DNA/MULE-NOF',
       'DNA/Crypton-H', 'DNA/hAT', 'DNA/Academ-1', 'DNA/IS3EU',
       'DNA/TcMar-Stowaway', 'DNA/Sola-3', 'DNA/TcMar-Tc2',
       'DNA/Kolobok-T2', 'DNA/hAT-hobo', 'DNA/TcMar-Pogo', 'DNA/Merlin',
       'LTR/Copia', 'DNA/MULE', 'LINE/CRE', 'DNA/TcMar-Tigger',
       'LINE/Rex-Babar', 'DNA/PIF-Spy', 'LINE/R2', 'DNA/hAT-hAT19',
       'DNA/TcMar-Sagan', 'Other/DNA_virus', 'DNA/Ginger-2'], dtype=object)


df = df[df["name"] != "Other/DNA_virus"]

# カテゴリ分け   
def assign_category(name):
    if "DNA" in name:
        return "DNA"
    elif "LINE" in name:
        return "LINE"
    elif "RC" in name:
        return "RC"
    elif "LTR" in name:
        return "LTR"
    elif "PLE" in name:
        return "PLE"
    elif "Satellite" in name:
        return "Satellite"
    elif "Simple_repeat" in name:
        return "Simple_repeat"
    elif "Low_complexity" in name:
        return "Low_complexity"
    elif "Unknown" in name:
        return "Unknown"

df["category"] = df["name"].apply(assign_category)

# 並び順
order = ["DNA", "RC", "LINE", "LTR", "PLE"]


#配色
base_palette = {"DNA": "#d94454", "RC": "#f2922c", "LINE": "#43b9a6", "LTR": "#53a550", "PLE": "#e7b540"}
palette = [base_palette[c] for c in order]

#violinplot
plt.figure(figsize=(8.4, 5.0))
ax = sns.violinplot(data=df, x="category", y="percent_m", order=order, inner=None, cut=0, width=0.8, linewidth=0.3, palette=palette, density_norm="count")
ax.set_xlabel("")
ax.set_ylabel("percent_m")
ax.spines[['top', 'right']].set_visible(False)
plt.tight_layout()
plt.savefig("violin_percent_m_by_TE_category.svg", format="svg", bbox_inches="tight")

#boxplot
plt.figure(figsize=(5.0, 5.0)) 
ax = sns.boxplot(data=df, x="category", y="percent_m", order=order, width=0.5, showfliers=False, palette=palette, boxprops=dict(edgecolor="black", linewidth=0.3), whiskerprops=dict(color="black", linewidth=0.3), medianprops=dict(color="black", linewidth=0.3))
ax.set_xlabel("")
ax.set_ylabel("percent_m")
ax.spines[['top', 'right']].set_visible(False)
plt.tight_layout()
plt.savefig("boxplot_percent_m_by_TE_category.svg", format="svg", bbox_inches="tight")

#メモ：カテゴリ別TEメチル化率#
summary = (df.groupby("category")["percent_m"].describe(percentiles=[0.25, 0.5, 0.75]).reset_index())
>>> summary
         category     count       mean        std  min       25%       50%        75%    max
0             DNA  597308.0  20.222756  33.491192  0.0  0.000000  1.464169  20.454546  100.0
1            LINE  310015.0  20.792097  33.953555  0.0  0.000000  1.443001  33.921842  100.0
2             LTR   10622.0  22.048095  34.470876  0.0  0.685824  1.560096  55.070170  100.0
3  Low_complexity   43356.0   2.464182  13.423839  0.0  0.000000  0.000000   0.000000  100.0
4             PLE   25550.0  16.447364  31.732795  0.0  0.000000  0.000000   4.347826  100.0
5              RC   19836.0  17.905027  31.442792  0.0  0.000000  1.369863   8.001067  100.0
6       Satellite   30619.0  30.310518  39.042579  0.0  0.000000  1.961722  78.961802  100.0
7   Simple_repeat  347283.0   2.218450  13.071044  0.0  0.000000  0.000000   0.000000  100.0
8         Unknown  504751.0  23.833136  35.959754  0.0  0.000000  1.612903  69.090910  100.0


#exon, intronのメチル化率
df1 = pd.read_csv("5mC_in_exon.tsv", sep="\t")
df1["name"] = "exon"
df2 = pd.read_csv("5mC_in_intron.tsv", sep="\t")
df2["name"] = "intron"
df = pd.concat([df1, df2], axis=0, ignore_index=True)

# 並び順
order = ["exon", "intron"]

#violinplot
plt.figure(figsize=(3, 5))
ax = sns.violinplot(data=df, x="name", y="percent_m", order=order, inner=None, cut=0, density_norm="count", width=0.8, linewidth=0.3)
ax.set_xlabel("")
ax.set_ylabel("percent_m")
ax.spines[['top', 'right']].set_visible(False)
plt.tight_layout()
plt.savefig("violin_percent_m_by_exon_intron.svg", format="svg", bbox_inches="tight")

#boxplot
plt.figure(figsize=(3, 5)) 
ax = sns.boxplot(data=df, x="name", y="percent_m", order=order, width=0.5, showfliers=False, boxprops=dict(edgecolor="black", linewidth=0.3), whiskerprops=dict(color="black", linewidth=0.3), medianprops=dict(color="black", linewidth=0.3))
ax.set_xlabel("")
ax.set_ylabel("percent_m")
ax.spines[['top', 'right']].set_visible(False)
plt.tight_layout()
plt.savefig("boxplot_percent_m_by_exon_intron.svg", format="svg", bbox_inches="tight")


#メモ：exon, intronメチル化率#
summary = (df.groupby("name")["percent_m"].describe(percentiles=[0.25, 0.5, 0.75]).reset_index())
>>> summary
     name     count       mean        std  min  25%        50%        75%    max
0    exon  208465.0  32.650439  38.362015  0.0  0.0   2.564103  78.333336  100.0
1  intron  177854.0  37.340849  38.030733  0.0  0.0  13.968653  79.896910  100.0


#promoter, gene, intergenicのメチル化率
df3 = pd.read_csv("5mC_in_promoter.tsv", sep="\t")
df3["name"] = "promoter"
df4 = pd.read_csv("5mC_in_gene.tsv", sep="\t")
df4["name"] = "gene"
df5 = pd.read_csv("5mC_in_intergenic.tsv", sep="\t")
df5["name"] = "intergenic"
df = pd.concat([df3, df4, df5], axis=0, ignore_index=True)

# 並び順
order = ["promoter", "intergenic", "gene"]

#violinplot
plt.figure(figsize=(4.5, 5))
ax = sns.violinplot(data=df, x="name", y="percent_m", order=order, inner=None, cut=0, density_norm="count", width=0.8, linewidth=0.3)
ax.set_xlabel("")
ax.set_ylabel("percent_m")
ax.spines[['top', 'right']].set_visible(False)
plt.tight_layout()
plt.savefig("violin_percent_m_by_gene_intergenic_promoter.svg", format="svg", bbox_inches="tight")

#boxplot
plt.figure(figsize=(4.5, 5)) 
ax = sns.boxplot(data=df, x="name", y="percent_m", order=order, width=0.5, showfliers=False, boxprops=dict(edgecolor="black", linewidth=0.3), whiskerprops=dict(color="black", linewidth=0.3), medianprops=dict(color="black", linewidth=0.3))
ax.set_xlabel("")
ax.set_ylabel("percent_m")
ax.spines[['top', 'right']].set_visible(False)
plt.tight_layout()
plt.savefig("boxplot_percent_m_by_gene_intergenic_promoter.svg", format="svg", bbox_inches="tight")


#メモ：exon, intronメチル化率#
summary = (df.groupby("name")["percent_m"].describe(percentiles=[0.25, 0.5, 0.75]).reset_index())
>>> summary
         name    count       mean        std  min       25%       50%        75%        max
0        gene  28515.0  28.514044  32.514399  0.0  1.596590  3.956835  64.974423   97.05882
1  intergenic  26222.0  14.214935  22.268003  0.0  1.702943  2.823317  16.703125  100.00000
2    promoter  22390.0  13.747067  26.287772  0.0  1.372998  1.940447   4.099936  100.00000
```

## 5.2 Methylation levels of conserved genes vs. other genes
```python
python3
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import mannwhitneyu
import pandas as pd
import re

file = pd.read_csv("5mC_in_gene.tsv", sep="\t")
expr = pd.read_csv("/data2/Hydra_RNA_seq_DATA_GeneBay_2022-01-45-01-1_2022-3-10/analysis/cont.genes.results", sep="\t")
expr["logTPM"] = np.log10(expr["TPM"] + 1)
#file_x_file_y_を削除する
expr["gene_id"] = expr["gene_id"].str.replace(r"^file_\d+_file_\d+_", "", regex=True)

# 保存された遺伝子リストの読み込み
#all_common_og_lists.txt：orthofinderの出力から33種が共通して保有するオルソグループのJNIG IDを記載したファイル
common_og_lists = pd.read_csv("all_common_og_lists.txt", header=None)[0].tolist() 
#file_x_file_y_を削除する
common_og_lists = [re.sub(r"^file_\d+_file_\d+_", "", x)  for x in common_og_lists]

# name と gene_id で結合
merged = pd.merge(file, expr[["gene_id", "logTPM"]], left_on="name", right_on="gene_id", how="left")
df = merged.dropna(subset=["percent_m", "logTPM"])
df["highlight"] = df["name"].isin(common_og_lists)

#メモ：conserved vs other genes でメチル化率は有意に異なるか
group1 = df.loc[df["highlight"] == True, "percent_m"]
group2 = df.loc[df["highlight"] == False, "percent_m"]
stat, p = mannwhitneyu(group1, group2, alternative='two-sided')
print(f"Mann–Whitney U test: U = {stat:.3f}, p = {p:.3e}")
#Mann–Whitney U test: U = 45259269.000, p = 0.000e+00

#boxplot
data = [group2, group1]
labels = ["Other genes", "Conserved genes"]
colors = ["#b5b6b6", "#920783"] 
plt.figure(figsize=(5,5))
bp = plt.boxplot(data, patch_artist=True, labels=labels, boxprops=dict(linewidth=0.3), medianprops=dict(color='black', linewidth=2), whiskerprops=dict(linewidth=0.3), capprops=dict(linewidth=0.3))
# box に色をつける
for patch, color in zip(bp['boxes'], colors):
    patch.set_facecolor(color)
plt.ylabel("Methylation level")
plt.tight_layout()
plt.savefig("Methylation_level_by_conserved_other.svg")

#メモ：conserved vs other genes で発現量は有意に異なるか#
group1 = df.loc[df["highlight"] == True, "logTPM"]
group2 = df.loc[df["highlight"] == False, "logTPM"]
stat, p = mannwhitneyu(group1, group2, alternative='two-sided')
print(f"Mann–Whitney U test: U = {stat:.3f}, p = {p:.3e}")
#Mann–Whitney U test: U = 49851200.500, p = 0.000e+00

#boxplot
data = [group2, group1]
labels = ["Other genes", "Conserved genes"]
colors = ["#b5b6b6", "#920783"] 
plt.figure(figsize=(3,5))
bp = plt.boxplot(data, patch_artist=True, labels=labels, boxprops=dict(linewidth=0.3), medianprops=dict(color='black', linewidth=2), whiskerprops=dict(linewidth=0.3), capprops=dict(linewidth=0.3))
# box に色をつける
for patch, color in zip(bp['boxes'], colors):
    patch.set_facecolor(color)
plt.ylabel("expression level")
plt.tight_layout()
plt.savefig("expression_level_by_conserved_other.svg")
```

##  5.3 Scatter plot of methylation level vs. gene expression
```python
#gene分のみ

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

file = pd.read_csv("5mC_in_gene.tsv", sep="\t")
file["name"] = file["name"].str.extract(r"gene_id=([^;]+)")
expr = pd.read_csv("/data2/Hydra_RNA_seq_DATA_GeneBay_2022-01-45-01-1_2022-3-10/analysis/cont.genes.results", sep="\t")
expr["logTPM"] = np.log10(expr["TPM"] + 1)
#file_x_file_y_を削除する
expr["gene_id"] = expr["gene_id"].str.replace(r"^file_\d+_file_\d+_", "", regex=True)

# 保存された遺伝子リストの読み込み
#all_common_og_lists.txt：orthofinderの出力から33種が共通して保有するオルソグループのJNIG IDを記載したファイル
common_og_lists = pd.read_csv("all_common_og_lists.txt", header=None)[0].tolist() 
#file_x_file_y_を削除する
common_og_lists = [re.sub(r"^file_\d+_file_\d+_", "", x)  for x in common_og_lists]

# name と gene_id で結合
merged = pd.merge(file, expr[["gene_id", "logTPM"]], left_on="name", right_on="gene_id", how="left")
df = merged.dropna(subset=["percent_m", "logTPM"])

#メモ：(conserved / nonconserved) x (hypermethyl / hypomethyl) 4群のCVを計算#
df["conserved_gene"] = df["name"].isin(common_og_lists).map({True: "conserved", False: "nonconserved"})
# percent_m で2群（閾値:メチル化率60%）
df["group_pm"] = np.where(df["percent_m"] >= 60, "hyper", "hypo")
# 4群を作成
df["group"] = df["group_pm"] + "_" + df["conserved_gene"]
# CV（SD / Mean）を計算
cv_df = (df.groupby("group")["logTPM"].agg(["mean", "std"]).assign(CV=lambda x: x["std"] / x["mean"]))
print(cv_df)
>>> print(cv_df)
                        mean       std        CV
group
hyper_conserved     1.449974  0.535543  0.369347
hyper_nonconserved  0.880028  0.670156  0.761517
hypo_conserved      1.314029  0.883477  0.672342
hypo_nonconserved   0.397404  0.620251  1.560758


#scatter plot（logTPM vs percent_m）
df["highlight"] = df["name"].isin(common_og_lists)
plt.figure(figsize=(6, 5))
plt.scatter(df.loc[~df["highlight"], "percent_m"],
            df.loc[~df["highlight"], "logTPM"],
            color="#b5b6b6", marker=".", s=5, label="other genes")
plt.scatter(df.loc[df["highlight"], "percent_m"],
            df.loc[df["highlight"], "logTPM"],
            color="#920783", marker=".", s=5, label="conserved genes")
plt.xlabel("percent_m (%)")
plt.ylabel("logTPM")
plt.tight_layout()
plt.savefig("percent_m_logTPM_correlation_conserved_other.svg")
```


